### ISSUE TITLE
Missing documentation for method defined in header during autocomplete

### ISSUE THREAD

**Issue by bevin-hansson** [clangd/clangd#2252]
(I'm unsure if this is clangd or vscode-clangd-specific, but I'm filing it here so someone who does know can triage it.)

When autocompleting, say, a method invocation, you get a list of items with the return type in a side box.

![auto1](https://github.com/user-attachments/assets/2c53e4f9-bbe0-49a7-8ed1-cae874f4be0f)

This is rather anemic. With no visible documentation, there's no way of knowing what the methods even do. It's only when you perform the autocomplete action that you get the documentation and a more comprehensive method signature with the return type:

![auto2](https://github.com/user-attachments/assets/2ee6780c-9a50-4484-b4db-8f2859c658a2)

This is the content I would like to see in the side box before I've hit Enter and performed the autocomplete, not after.

I guess this is a bit similar to #80, but I want it for the autocomplete, not the hover.

**Comment by HighCommander4** [comment]
#564 is a similar report about fields.

**Comment by HighCommander4** [comment]
Explanation for why this doesn't work: https://github.com/llvm/llvm-project/commit/be0eb8f498e89aede0083a082f3f455783af5d91

**Comment by bevin-hansson** [comment]
I see, interesting. I didn't realize there was already a ticket about it. Sorry about that.

This feels like a pretty useful feature to be missing for so long. I'm also not sure I understand why this works for functions in headers but not for members, or why it works for the signatureHelp action. The latter is because it's strictly related to code completion, I guess, but why does the functionality of that endpoint not suffer from the same underlying issue?

**Comment by bevin-hansson** [comment]
An initial attempt to fix it seems to have resolved the basic reproducer. I added the following to CodeCompletionBuilder::CodeCompletionBuilder at the end of `if (C.SemaResult)`:
```
  if (!Completion.Documentation &&
      C.SemaResult->Kind == CodeCompletionResult::RK_Declaration) {
    auto ID = getSymbolID(C.SemaResult->getDeclaration());
    if (ID) {
      LookupRequest Req;
      Req.IDs.insert(ID);
      Opts.Index->lookup(Req, [&](const Symbol &S) {
        if (!S.Documentation.empty()) {
          Completion.Documentation.emplace();
          parseDocumentation(formatDocumentation(*SemaCCS, S.Documentation), *Completion.Documentation);
        }
      });
    }
  }
```
If the SemaResult doesn't have any documentation, look up the declaration in the index instead and see if there's something there. I'm not terribly familiar with clangd so I don't know if this is the right way to do this.

**Comment by HighCommander4** [comment]
@bevin-hansson Yes, I think that's the way to fix it! I believe this is what the author of https://github.com/llvm/llvm-project/commit/be0eb8f498e89aede0083a082f3f455783af5d91 meant by "rely on the index instead", and this was subsequently implemented for signature help in https://github.com/llvm/llvm-project/commit/8a0f76b855afdf1099e4447dd30543ea7cb9c562, but not for code complete.

One tweak I think we'll want to make to what you wrote is, for performance reasons, we'll want to batch the index lookups together into a single request for all decls that we're returning, the way the [signature help codepath](https://searchfox.org/llvm/rev/b7e75a76449e80d8b6caa1f447d536ffa231783a/clang-tools-extra/clangd/CodeComplete.cpp#1113-1130) does.

Are you interested in putting together a PR?

**Comment by HighCommander4** [comment]
> I'm also not sure I understand why this works for functions in headers but not for members,

For namespace-scope functions, we don't use Sema at all and rely entirely on the index, and so we pick up documentation from the index around [here](https://searchfox.org/llvm/rev/b7e75a76449e80d8b6caa1f447d536ffa231783a/clang-tools-extra/clangd/CodeComplete.cpp#507).

(The reason we avoid Sema for namespace-scope functions is that the AST for much of the translation unit (the "preamble", ~ the contents of all included headers) is stored in serialized form, and doing a namespace-scope lookup would require deserializing all decls in that namespace, which could we quite slow for a large namespace like `std` or the global namespace; it would be linear time in the number of declarations in the namespace.)

> or why it works for the signatureHelp action

(This is probably clear by now, but that's because of the mentioned https://github.com/llvm/llvm-project/commit/8a0f76b855afdf1099e4447dd30543ea7cb9c562 adding an index lookup for signature help.)

**Comment by bevin-hansson** [comment]
> One tweak I think we'll want to make to what you wrote is, for performance reasons, we'll want to batch the index lookups together into a single request for all decls that we're returning, the way the [signature help codepath](https://searchfox.org/llvm/rev/b7e75a76449e80d8b6caa1f447d536ffa231783a/clang-tools-extra/clangd/CodeComplete.cpp#1113-1130) does.
> 
> Are you interested in putting together a PR?

I can make these changes and put together something. I'm not sure about tests, though.
